### ISSUE TITLE
[Bug] `get_dynamic_side_mask()` Ignores Pitch Angle — Large Pitch Face Swap Produces Visible Artifacts

### ISSUE THREAD

**Issue by Umstead** [VisoMasterFusion/VisoMaster-Fusion#254]
## Summary

`get_dynamic_side_mask()` in `frame_worker.py` receives `pitch_deg` as a parameter but **completely ignores it**. When a target face has a large pitch angle (looking up or down >30°), no vertical gradient masking is applied, causing visible blending artifacts at the forehead (pitch up) or chin/neck (pitch down) boundary.

**Affected file**: `app/processors/workers/frame_worker.py`, line 3777–3834  
**Related file**: `app/processors/utils/faceutil.py`, line 438–452 (`align_crop`)

---

### Root Cause

The function signature accepts `pitch_deg`:

```python
# frame_worker.py:3777
def get_dynamic_side_mask(
    self, yaw_deg, pitch_deg, height, width, device, parameters, kps_5, tform
):
```

And the caller correctly computes and passes it:

```python
# frame_worker.py:4330-4340
yaw_deg, pitch_deg = faceutil.calc_face_yaw_pitch(kps_5)
side_mask = self.get_dynamic_side_mask(
    yaw_deg,
    pitch_deg,    # ← passed in
    ...
)
```

But the function body **only uses `yaw_deg`** — `pitch_deg` is never referenced:

```python
# frame_worker.py:3805-3832
abs_yaw = abs(yaw_deg)
if abs_yaw > start_angle:
    # ... horizontal gradient masking based on yaw ...
    # pitch_deg is NEVER used
```

### Symptoms

| Pitch Angle | Symptom |
|-------------|---------|
| **0° – 15°** (frontal) | No visible issue (pitch masking not needed) |
| **15° – 30°** (slight up/down) | Mild blending artifacts at forehead or chin edge |
| **30° – 45°** (moderate) | Noticeable hard edge / halo at forehead (looking up) or chin/neck (looking down) |
| **>45°** (extreme) | Severe artifacts — swapped face appears to "float" or has a visible seam |

### Why It Matters

The yaw-based horizontal masking works well for profile faces (left/right rotation). But large pitch angles cause a **different kind of occlusion problem**:

- **Looking up**: The forehead and top-of-head region becomes unreliable — the face parser may include non-face pixels (hair, background sky) in the mask.
- **Looking down**: The chin and neck region becomes unreliable — the mask may extend below the jawline into the neck.

Without pitch-aware masking, these unreliable regions are blended at full strength, creating visible artifacts.

---

### Problem

The alignment system uses 7 fixed templates (`src1`–`src7` in `faceutil.py`):

| Template | Purpose |
|----------|---------|
| `src1`–`src5` | Yaw angles (left profile → right profile) |
| `src6` | **Pitch up** (fixed) |
| `src7` | **Pitch down** (fixed) |

`get_matrix()` (line 423) selects the best template by minimizing alignment error. However, `src6` and `src7` represent **extreme** pitch angles (~45–60°). For intermediate angles (20°, 30°, 40°), the system must choose between:

1. The frontal template (`src3`) — poor match for moderate pitch
2. The extreme pitch template (`src6`/`src7`) — over-correction for moderate pitch

There is **no interpolation** between templates for intermediate angles.

### Impact

For a face at pitch 25°:
- `src3` (frontal) error: moderate — eyes/nose/mouth vertical spacing doesn't match
- `src6` (extreme pitch up) error: moderate — over-shifts the landmarks
- `get_matrix()` picks whichever has lower error, but both are suboptimal

---

### Fix 1: Pitch-Aware Vertical Masking (frame_worker.py)

Add vertical gradient masking in `get_dynamic_side_mask()`, mirroring the existing yaw logic:

```python
# --- Pitch-based vertical masking ---
abs_pitch = abs(pitch_deg)
if abs_pitch > start_angle:
    pitch_excess = max(0, abs_pitch - start_angle)
    strength_pitch = min(pitch_excess / 45.0, 1.0) * max_strength
    linspace_y = torch.linspace(0, 1, height, device=device).view(1, height, 1)

    if pitch_deg > 0:
        # Looking Up -> Fade the TOP (forehead unreliable)
        fade_end = 0.35
        grad_pitch = torch.clamp(linspace_y / fade_end, 0, 1)
        grad_pitch = 1.0 - (1.0 - grad_pitch) * strength_pitch
        mask = mask * grad_pitch
    else:
        # Looking Down -> Fade the BOTTOM (chin/neck unreliable)
        fade_start = 0.65
        grad_pitch = torch.clamp(
            (linspace_y - fade_start) / (1.0 - fade_start), 0, 1
        )
        grad_pitch = 1.0 - grad_pitch
        mask = mask * grad_pitch
```

**Reuses existing UI controls**: `ProfileAngleMaskThresholdSlider` and `ProfileAngleMaskStrengthSlider` — no new settings needed.

### Fix 2: Adaptive Pitch Template Interpolation (faceutil.py)

Add a function that smoothly interpolates between the frontal template and the pitch template based on the actual pitch angle:

```python
def get_adaptive_pitch_template(pitch_deg, base_template):
    """Interpolate between frontal (src3) and pitch (src6/src7) templates."""
    frontal = base_template[2].copy()
    if abs(pitch_deg) < 15.0:
        return frontal
    clamped = min(abs(pitch_deg), 75.0)
    pitch_template = base_template[5] if pitch_deg > 0 else base_template[6]
    t = (clamped - 15.0) / 60.0
    t = t * t * (3.0 - 2.0 * t)  # smoothstep
    return (frontal * (1.0 - t) + pitch_template * t).astype(np.float32)
```

Then use it in `align_crop()` when `mode="arcfacemap"`:

```python
if mode == "arcfacemap":
    templates = float(image_size) / 112.0 * src_map[112]
    try:
        _, pitch = calc_face_yaw_pitch(lmk)
        matrix = get_matrix_with_pitch(lmk, templates, pitch_deg=pitch)
    except Exception:
        matrix = get_matrix(lmk, templates)
```

---

## Files to Modify

| File | Change | Lines |
|------|--------|-------|
| `app/processors/workers/frame_worker.py` | Add pitch-based vertical masking in `get_dynamic_side_mask()` | ~3834 (after yaw block) |
| `app/processors/utils/faceutil.py` | Add `get_adaptive_pitch_template()` and `get_matrix_with_pitch()` | After line 104 |
| `app/processors/utils/faceutil.py` | Modify `align_crop()` to use pitch-adaptive template | Line 452 |

---

## Risk Assessment

| Risk | Level | Mitigation |
|------|-------|------------|
| Regression in frontal faces | Low | Pitch masking only triggers above threshold (same as yaw) |
| Performance impact | Negligible | One extra `torch.linspace` + `clamp` per frame |
| Interaction with yaw masking | Low | Masks are multiplied — both dimensions handled independently |
| Template interpolation accuracy | Low | Fallback to standard `get_matrix()` on any error |

---

**Comment by Elricfae** [comment]
As i first added this it did use the pitch for masking. But after some testing it didn't give good result and it was deactivated, the code base for it is still present though.
Not sure i'll take another look at it.
